### PR TITLE
Stop the running log backup when deleting CR

### PR DESCRIPTION
**Pull request by RidRisR**
### What problem does this PR solve?

In the previous operator version, delete backup CR will do no clean up on running log backup.
Users may mis-operator and find it hard to running task.

### What is changed and how does it work?

1. Add enum for cleanPolicy, make input standardized.
   Now, only `Retain;OnFailure;Delete` are accepted as cleanPolicy input, and default is `Retain`.
   (This is actually no change to user behavior.)

2. Add finalizer and clean logic check for on-going log backup
    All log backup will have finalizer now. Operator will check if the log backup "on track". If so, a new task to stop log backup will run.
    The "on track" state is `BackupScheduled || BackupPrepare || BackupRunning || BackupPaused || BackupStop`. We lack the ability to check whether the log backup should be stopped if it failed. If the task is on fail states, we do not provide auto cleaning.

### Code changes

- [ ] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

1. Add enum for cleanPolicy, make input standardized.
   Now, only Retain;OnFailure;Delete are accepted as cleanPolicy input, and Retain is default.
   This is actually no change to user behavior.
2. Add finalizer and clean logic check for on-going log backup
    All log backup will have finalizer now. Operator will check if the log backup "on track". If so, a new task to stop log backup will run.
    The "on track" state is `BackupScheduled || BackupPrepare || BackupRunning || BackupPaused`. 
    We lack the ability to check whether the log backup should be stopped if it failed. If the task is on states, we do not provide auto cleaning.
```

### CODE REVIEW

File: pkg/backup/backup/backup_cleaner.go [hunk 1]
```diff
@@ -52,7 +53,73 @@ func NewBackupCleaner(deps *controller.Dependencies, statusUpdater controller.Ba
...
+	ns := backup.GetNamespace()
+	name := backup.GetName()
+	stopLogJobName := backup.GetCleanLogBackupJobName()
+
+	var err error
+	_, err = bc.deps.JobLister.Jobs(ns).Get(stopLogJobName)
+	if err == nil {
+		// already have a clean job running，return directly
+		return nil
+	}
```
Comment (csuzhangxc) [inline pkg/backup/backup/backup_cleaner.go:91]: do we need to differ the NotFound error with other errors?

File: pkg/backup/backup/backup_cleaner.go [hunk 2]
```diff
@@ -218,6 +285,205 @@ func (bc *backupCleaner) makeCleanJob(backup *v1alpha1.Backup) (*batchv1.Job, st
 	return job, "", nil
 }
 
+func (bc *backupCleaner) makeStopLogBackupJob(backup *v1alpha1.Backup) (*batchv1.Job, string, error) {
```
Comment (csuzhangxc) [inline pkg/backup/backup/backup_cleaner.go:288]: is it possible to re-use `makeBRBackupJob`?

Comment (RidRisR) [reply pkg/backup/backup/backup_cleaner.go:288]: The key problem is they have different caller. backupCleaner can't call method of backupManager, and it is weird to use backupCleaner to build other tasks.

File: pkg/backup/backup/backup_cleaner.go (hunk 1 above)
Comment (RidRisR) [reply pkg/backup/backup/backup_cleaner.go]: Yes

File: pkg/backup/backup/backup_cleaner.go [hunk 3]
```diff
@@ -52,7 +53,81 @@ func NewBackupCleaner(deps *controller.Dependencies, statusUpdater controller.Ba
...
+	}
+
+	klog.Infof("start to clean backup %s/%s", backup.GetNamespace(), backup.GetName())
+	if err := bc.CleanLogBackup(backup); err != nil {
+		return err
+	}
+	return bc.CleanData(backup)
+}
+
+func (bc *backupCleaner) CleanLogBackup(backup *v1alpha1.Backup) error {
```
Comment (WangLe1321) [inline pkg/backup/backup/backup_cleaner.go:68]: it seems `StopLogBackup` better?

File: pkg/backup/backup/backup_cleaner.go [hunk 4]
```diff
@@ -52,7 +53,81 @@ func NewBackupCleaner(deps *controller.Dependencies, statusUpdater controller.Ba
...
+
+	return bc.statusUpdater.Update(backup, &v1alpha1.BackupCondition{
+		Command: v1alpha1.LogStopCommand,
+		Type:    v1alpha1.BackupScheduled,
+		Status:  corev1.ConditionTrue,
+	}, nil)
+}
+
+func (bc *backupCleaner) CleanData(backup *v1alpha1.Backup) error {
+	if backup.DeletionTimestamp == nil || !v1alpha1.IsCleanCandidate(backup) || v1alpha1.NeedRetainData(backup) {
```
Comment (WangLe1321) [inline pkg/backup/backup/backup_cleaner.go:130]: For the log backup, we should ensure it is stopped before clean data.

File: pkg/backup/backup/backup_cleaner.go (hunk 2 above)
Comment (WangLe1321) [reply pkg/backup/backup/backup_cleaner.go:296]: How about reuse `makeCleanJob`? Their main difference is parameter.

Comment (RidRisR) [reply pkg/backup/backup/backup_cleaner.go:296]: I prefer leave it be, they are actually using different pods.
I tried to combine them into one, but that only makes the function fat and mess.

File: pkg/backup/backup/backup_cleaner.go (hunk 3 above)
Comment (RidRisR) [reply pkg/backup/backup/backup_cleaner.go:68]: Ok, no diffence to me.

File: pkg/backup/backup/backup_cleaner.go (hunk 4 above)
Comment (RidRisR) [reply pkg/backup/backup/backup_cleaner.go:130]: Right

File: pkg/controller/backup_status_updater.go [hunk 5]
```diff
@@ -377,9 +377,12 @@ func updateWholeLogBackupStatus(backup *v1alpha1.Backup, condition *v1alpha1.Bac
...
 	}
 
-	// DeletionTimestamp is not nil when delete and clean backup, condition and status can be directly used to update whole status.
+	// restruct status and condition according to subcommand's detail info, and they wiil be used to update whole log backup status.
+	newStatus := restructStatus()
+	newCondition := restructCondition()
+
 	if backup.DeletionTimestamp != nil {
-		return doUpdateStatusAndCondition(condition, status)
+		return doUpdateStatusAndCondition(newCondition, status)
```
[the remark this answers is not in the document]
Comment (RidRisR) [reply pkg/controller/backup_status_updater.go:385]: We could, but not necessary.

File: tests/e2e/br/br.go [hunk 6]
```diff
@@ -1341,7 +1444,7 @@ func deleteBackup(f *e2eframework.Framework, name string) error {
 		return err
 	}
 
-	if err := brutil.WaitForBackupDeleted(f.ExtClient, ns, name, time.Second*60); err != nil {
+	if err := brutil.WaitForBackupDeleted(f.ExtClient, ns, name, time.Second*60*15); err != nil {
```
[the remark this answers is not in the document]
Comment (RidRisR) [reply tests/e2e/br/br.go:1447]: It might timeout for 60s. 15 min is just a arbitrary time that long enough.

**Review by WangLe1321 — APPROVED** [review]

**Review by csuzhangxc — APPROVED** [review]

File: pkg/backup/backup/backup_cleaner.go [hunk 7]
```diff
@@ -218,60 +293,311 @@ func (bc *backupCleaner) makeCleanJob(backup *v1alpha1.Backup) (*batchv1.Job, st
...
 
-	// log backup may have multiple jobs
-	backupJobNames := make([]string, 0)
-	if backup.Spec.Mode == v1alpha1.BackupModeLog {
-		backupJobNames = append(backupJobNames, backup.GetAllLogBackupJobName()...)
-	} else {
-		backupJobNames = append(backupJobNames, backup.GetBackupJobName())
+	tc, err := bc.deps.TiDBClusterLister.TidbClusters(backupNamespace).Get(backup.Spec.BR.Cluster)
+	if err != nil {
+		return nil, fmt.Sprintf("failed to fetch tidbcluster %s/%s", backupNamespace, backup.Spec.BR.Cluster), err
```
Comment (aravndr) [inline pkg/backup/backup/backup_cleaner.go:307]: What if the cluster has already been deleted? how does one clean up the backups then?

**Pull request merged**